tracer/src/Datadog.Trace.Coverage.collector/CoverageCollector.cs

```diff
@@ -0,0 +1,253 @@
+                    return;
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.Error(ex);
+                _ciVisibilitySettings = null;
+            }
+
+            if (_events is not null)
+            {
+                _events.SessionStart += OnSessionStart;
+                _events.SessionEnd += OnSessionEnd;
+            }
+        }
+
+        private void OnSessionStart(object? sender, SessionStartEventArgs e)
+        {
+            _logger?.SetContext(e.Context);
+            var testSources = e.GetPropertyValue("TestSources");
+            if (testSources is string testSourceString)
+            {
+                // Process folder
+                var outputFolder = Path.GetDirectoryName(testSourceString);
+                if (outputFolder is not null)
+                {
+                    BackupFolder(outputFolder);
+                    ProcessFolder(outputFolder, SearchOption.TopDirectoryOnly);
+                }
+            }
+            else if (testSources is List<string> testSourcesList)
+            {
+                // Process folder
+                foreach (var source in testSourcesList)
+                {
+                    var outputFolder = Path.GetDirectoryName(source);
+                    if (outputFolder is not null)
+                    {
+                        BackupFolder(outputFolder);
+                        ProcessFolder(outputFolder, SearchOption.TopDirectoryOnly);
+                    }
+                }
+            }
+            else
+            {
+                // Process folder
+                BackupFolder(Environment.CurrentDirectory);
+                ProcessFolder(Environment.CurrentDirectory, SearchOption.AllDirectories);
+            }
+        }
+
+        private void OnSessionEnd(object? sender, SessionEndEventArgs e)
+        {
+            _logger?.SetContext(e.Context);
+        }
+
+        private void ProcessFolder(string folder, SearchOption searchOption)
+        {
+            if (_tracerHome is null)
+            {
+                return;
+            }
+
+            var processedDirectories = new HashSet<string>();
+            var numAssemblies = 0;
+            var tracerAssemblyName = typeof(Tracer).Assembly.GetName().Name;
+
+            // Process assemblies in parallel.
+            Parallel.ForEach(
+                Directory.EnumerateFiles(folder, "*.*", searchOption),
+                file =>
+                {
+                    // Skip the Datadog.Trace assembly
+                    if (tracerAssemblyName == Path.GetFileNameWithoutExtension(file))
+                    {
+                        return;
+                    }
+
+                    var extension = Path.GetExtension(file).ToLowerInvariant();
+                    if (extension is ".dll" or ".exe")
```

Comment:
On linux, I think when published as a single executable, we _might not have a file extension_? 🤔 

Have you tried this with self-contained executables and/or single file executables? 

Can't imagine it works, or anyone would do it, just more thinking out loud 😄

Reply:
>On linux, I think when published as a single executable, we might not have a file extension? 

You're right, good catch.

> Have you tried this with self-contained executables and/or single file executables?

I haven't but I cannot think in a test project/suite scenario with self-contained assemblies.

> Can't imagine it works, or anyone would do it, just more thinking out loud

Well supporting this will depend on Mono.Cecil support, so what I can do is to test it (really don't know how to create a self-contained executable for a test project) and document it. I prefer to do this later in a different PR.